db/utils.js

```diff
@@ -0,0 +1,99 @@
+"use strict";
+
+const Breach = require("./models/Breach");
+const EmailHash = require("./models/EmailHash");
+
+const getSha1 = require("../sha1-utils");
+
+const DBUtils = {
+  async createBreach(name, meta) {
+    try {
+    return await Breach
+      .query()
+      .insert({ name, meta });
+    } catch(e) {}
+  },
+
+  async deleteBreach(id) {
+    await Breach.query().deleteById(id);
+  },
+
+  // Used internally, ideally should not be called by consumers.
+  async _addEmailHash(sha1, email) {
+    // Check if an entry exists
+    const existingEntries = await EmailHash
+      .query()
+      .where("sha1", sha1);
+
+    // If not, add it and return.
+    if (!existingEntries.length) {
+      return await EmailHash
+        .query()
+        .insert({ sha1, email });
+    }
+
+    // Entry existed, patch the email value if supplied.
+    if (email) {
+      return await existingEntries[0]
+        .$query()
+        .patch({ email })
+        .returning('*'); // Postgres trick to return the updated row as model.
+    }
+
+    return existingEntries[0];
+  },
+
+  async addSubscriber(email) {
+    const sha1 = getSha1(email);
+    return await this._addEmailHash(sha1, email);
+  },
+
+  async removeSubscriber(email) {
+    const sha1 = getSha1(email);
+
+    // Check if an entry exists.
+    const existingEntries = await EmailHash
+      .query()
+      .where("sha1", sha1);
+
+    // If not, nothing to be done, return.
+    if (!existingEntries.length) {
+      return;
+    }
+
+    // Patch out the email from the entry.
+    await existingEntries[0]
+      .$query()
+      .patch({ email: null })
+      .returning('*'); // Postgres trick to return the updated row as model.
+  },
+
+  async addBreachedHash(breachName, sha1) {
```

Comment:
A future refactor might be to use [Objection `upsertGraph` query](https://vincit.github.io/objection.js/#graph-upserts) for this?